core/shared/src/main/scala/fs2/internal/FreeC.scala

```diff
@@ -12,138 +12,204 @@ import scala.util.control.NonFatal
+      case r @ Result.Pure(_)           => Some(r: Result[R])
+      case r @ Result.Fail(_)           => Some(r: Result[R])
+      case r @ Result.Interrupted(_, _) => Some(r: Result[R])
+      case _                            => None
+    }
+
+    final case class Pure[F[_], R](r: R) extends FreeC[F, R] with Result[R] {
+      override def translate[G[_]](f: F ~> G): FreeC[G, R] =
+        this.asInstanceOf[FreeC[G, R]]
+      override def toString: String = s"FreeC.Pure($r)"
+    }
+
+    final case class Fail[F[_], R](error: Throwable) extends FreeC[F, R] with Result[R] {
+      override def translate[G[_]](f: F ~> G): FreeC[G, R] =
+        this.asInstanceOf[FreeC[G, R]]
+      override def toString: String = s"FreeC.Fail($error)"
+    }
+
+    // finalizerErrors contains accumulated errors during finalization of interruption.
+    // they will be presented to freeC after interupted scope will be closed
+    final case class Interrupted[F[_], R](scope: Token, finalizerErrors: Option[Throwable])
+        extends FreeC[F, R]
+        with Result[R] {
+      override def translate[G[_]](f: F ~> G): FreeC[G, R] =
+        this.asInstanceOf[FreeC[G, R]]
+      override def toString: String = s"FreeC.Interrupted($scope, ${})"
+    }
+
+    implicit class ResultSyntax[R](val self: Result[R]) extends AnyVal {
+
+      def map[R2](f: R => R2): Result[R2] = self match {
+        case Result.Pure(r) =>
+          try { Result.Pure(f(r)) } catch { case NonFatal(err) => Result.Fail(err) }
+        case failure @ Result.Fail(_)               => failure.asInstanceOf[Result[R2]]
+        case interrupted @ Result.Interrupted(_, _) => interrupted.asInstanceOf[Result[R2]]
+      }
+
+    }
+
   }
+
   final case class Eval[F[_], R](fr: F[R]) extends FreeC[F, R] {
     override def translate[G[_]](f: F ~> G): FreeC[G, R] =
       try Eval(f(fr))
-      catch { case NonFatal(t) => Fail[G, R](t) }
+      catch { case NonFatal(t) => Result.Fail[G, R](t) }
     override def toString: String = s"FreeC.Eval($fr)"
   }
-  final case class Bind[F[_], X, R](fx: FreeC[F, X], f: Either[Throwable, X] => FreeC[F, R])
+  final case class Bind[F[_], X, R](fx: FreeC[F, X], f: Result[X] => FreeC[F, R])
       extends FreeC[F, R] {
     override def toString: String = s"FreeC.Bind($fx, $f)"
   }
-  final case class Fail[F[_], R](error: Throwable) extends FreeC[F, R] {
-    override def translate[G[_]](f: F ~> G): FreeC[G, R] =
-      this.asInstanceOf[FreeC[G, R]]
-    override def toString: String = s"FreeC.Fail($error)"
-  }
-
-  private val pureContinuation_ = (e: Either[Throwable, Any]) =>
-    e match {
-      case Right(r) => Pure[Any, Any](r)
-      case Left(e)  => Fail[Any, Any](e)
-  }
 
-  def pureContinuation[F[_], R]: Either[Throwable, R] => FreeC[F, R] =
-    pureContinuation_.asInstanceOf[Either[Throwable, R] => FreeC[F, R]]
+  def pureContinuation[F[_], R]: Result[R] => FreeC[F, R] =
+    _.asFreeC[F]
```

Comment:
Don't want to create one instance here and share, like we did before?

Reply:
do you think this won't get inlined? I can ofc get instance back.

Reply:
I gave up predicting what gets inlined a long time ago. Seems like it would be a Function1 allocation but who knows.

Reply:
ok :-) will then reintroduce that val

Reply:
Ok I think I know why I have decided to not have val there anymore: 

former code 
```
private val pureContinuation_ = (e: Either[Throwable, Any]) =>
    e match {
      case Right(r) => Pure[Any, Any](r)
      case Left(e)  => Fail[Any, Any](e)
  }

  def pureContinuation[F[_], R]: Either[Throwable, R] => FreeC[F, R] =
    pureContinuation_.asInstanceOf[Either[Throwable, R] => FreeC[F, R]]

```
As you see there is anyhow typeCat essentially what _.asFreeC does :-) 
One of the advantages we can pass along Result to bind